README.md

```diff
@@ -1,55 +1,385 @@
+# test ()
+- : int * int list =
+(20, [2; 2; 2; 2; 2; 2; 2; 2; 2; 2; 1; 1; 1; 1; 1; 1; 1; 1; 1; 1])
+```
+
+Everything seems fine, right ? Except, it is not running in parallel as we can
+see from the consecutive `2` (pushed by `domainB`) and `1` pushed by `domainA`.
+This is because spawning a domain takes way more time than executing `work`, so
+`domainA` is long finished before `domainB` is even spawned. One way to overpass
+this issue is to increase the amount of work done in `work` (for example, by
+pushing more elements). Another way is to make sure the domains wait for each
+other before beginning their workload.
+
+We use a basic
+[barrier implementation](https://github.com/ocaml-multicore/saturn/blob/main/test/barrier/barrier.mli)
+to do that. Each domain will now wait for the other to reach the barrier before
+beginning pushing.
+
+```ocaml
+let work_par id barrier q =
+  Barrier.await barrier;
+  for i = 0 to 9 do
+    push q id
+  done
+```
+
+The `test` function is now:
+
+```ocaml
+let test_par () =
+  let barrier = Barrier.create 2 in
+  let q = create () in
+  let domainA = Domain.spawn (fun () -> work_par 1 barrier q) in
+  let domainB = Domain.spawn (fun () -> work_par 2 barrier q) in
+  Domain.join domainA;
+  Domain.join domainB;
+  (List.length !q, !q)
+```
+
+Let's run it:
+
+```ocaml
+# test_par ();;
+- : int * int list =
+(18, [2; 1; 2; 1; 2; 1; 2; 1; 2; 1; 2; 2; 1; 1; 1; 1; 2; 1])
+```
+
+Now, the `1` and the `2` are interleaved: domains are running in parallel. The
+resulting queue however contains only `18` elements whereas `20` were pushed.
+This is because we not only have a
+[race condition](https://en.wikipedia.org/wiki/Race_condition) here but also
+because `push` is a non-atomic operation. It requires to first read the content
+of the queue (`!q`) then write in it (`q := ...`). So, for example, when two
+domains try to push in parallel in an empty queue the following sequence can
+happen :
+
+- domain A read the queue : it is empty
+- domain B read the queue : it is still empty
+- domain A pushes `1` on the empty queue it has read before
+- domain B pushes `2` on the empty queue it has read before This sequence
+  results in a queue containing only one element of value `2`: the element
+  pushed by A is lost because B did not see it.
+
+This is a very common issue in parallel programming that can be avoided or
+bypassed in multiple ways, mainly:
+
+- use [`Atomic`](https://v2.ocaml.org/api/Atomic.html) for mutable variables,
+- use a mutual exclusion mechanism like
+  [`Mutex`](https://v2.ocaml.org/api/Mutex.html).
+
+However both solutions have their limits. Using mutexes or locks open the way to
+deadlock, livelock, priority inversion etc.. and often considerably limit the
+gain in performance of using multiple cores as it limits the part of the code
+effectively running in parallel. In another hand, atomics are -without a complex
+algorithm to combine them- only a solution for a single shared variable.
```

Comment:
```suggestion
performance gained by using multiple cores as they limit the parts of the code
effectively running in parallel. On the other hand, atomics are – without a complex
algorithm to combine them – only a solution for a single shared variable.
```

Reply:
Consider taking the spaces out around the m-dashes (if they'll indeed show up as m-dashes and not hyphens when live) or just put that aside in parentheses.

Reply:
Seems like British English prefers the en dash with spaces either side, (these are en dashes on my keyboard) not sure how this relates to Markdown though.